cloudmarker/alerts/emailalert.py

```diff
@@ -56,22 +62,58 @@ def done(self):
         this example implementation, we properly terminate the JSON
         array in the email body.
 
+        The connection is made based whether the SMTP or SMTP_SSL is
+        required which is determined by ``use_ssl`` param. The SMTP
+        login process is conducted if ``username`` param is provided.
+
         """
-        message = MIMEMultipart()
-        message['Date'] = email.utils.formatdate(localtime=True)
-        message['Subject'] = self.subject
-        message['From'] = self.sender
-        message['To'] = email.utils.COMMASPACE.join(self.to)
+        _message = MIMEMultipart()
+        _message['Date'] = email.utils.formatdate(localtime=True)
+        _message['Subject'] = self._subject
+        _message['From'] = self._sender
+        _message['To'] = email.utils.COMMASPACE.join(self._to)
 
         # Incase if the buffer is empty then the default message will be sent
-        if self.stringbuffer:
-            self.body = ''.join(self.stringbuffer)
-        message.attach(MIMEText(self.body))
+        if self._stringbuffer:
+            self._body = ''.join(self._stringbuffer)
+        _message.attach(MIMEText(self._body))
 
-        smtp_connection = smtplib.SMTP(host=self.host, port=self.port)
+        _smtp = self._prepare_smtp_session()
         try:
-            smtp_connection.sendmail(self.sender, self.to, message.as_string())
+            _smtp.sendmail(self._sender, self._to, _message.as_string())
         except smtplib.SMTPException as e:
             _log.error('Failed to send email: %s', e)
         finally:
-            smtp_connection.quit()
+            _smtp.quit()
+
+    def _prepare_smtp_session(self):
+        """Return SMTP connection object.
+
+        Create a SMTP connection based on whether the SSL param.
+        If SSL connection is required (for eg: in case of gmail) then
+        method will return a SMTP_SSL connection. In other cases if the
+        authentication is not required or SMTP_SSL connection is not
+        required then a plain SMTP connection object is returned.
+        For gmail follow the steps given below even to work for 2 factor
+        authentication
+        1. Log-in into Gmail with your account
+        2. Navigate to https://security.google.com/settings/security/
+        apppasswords
+        3. In 'select app' choose 'custom', give it an arbitrary name and
+        press generate
+        4. It will give you 16 chars token.
+        5. Use that token as password for login
+        6. Host is smtp.gmail.com and port is 465
+
+        """
+        _smtp = None
+        if self._use_ssl:
+            _smtp = smtplib.SMTP_SSL(host=self._host, port=self._port)
+        else:
+            _smtp = smtplib.SMTP(host=self._host, port=self._port)
+            _smtp.starttls()
```

Comment:
The `smtp.starttls()` call must go after `smtp.set_debuglevel(True)` call otherwise. With the current pull request, we will get no clue about why a session fails if the session fails during `smtp.starttls()`. That partially defeats the purpose of having `smtp.set_debuglevel(True)`.

I have shared two code examples in a previous review to show how to write this code this correctly. Please see [this review comment](https://github.com/cloudmarker/cloudmarker/pull/75#discussion_r276239226).

Reply:
https://github.com/cloudmarker/cloudmarker/pull/75/commits/d092d76434f27c46450ee99c9fa9ed8ce75f0ba0#diff-28d4faa43f2d620f5a2c54123370ee47R109